### DISCUSSION TITLE
How to change to a page when a trigger occurs

### DISCUSSION THREAD

**Discussion by eboston** [HASwitchPlate/HASPone#26] [Q&A]
I currently have a button on a page that gets it text set based on an external trigger.  What happens with the current automation is it saves the currently active page, changes to page 1 and sets the text.  When button 4 is pressed, it resets the text and changes back to the original page.  And to make it even more complicated, I need to be able to set the times when the trigger can happen.

Is this something that can be done with existing blueprints or am I looking at needed to create my own?

Here is the current automation that I am using.  It is tied to a sensor in my mailbox that triggers the action.

`
########################################################################################     

  - alias: hasp_devplate_p1_SetMailTrigger
    trigger:
    - platform: mqtt
      topic: "mailbox/DoorTrigger"
    condition:
      condition: and
      conditions:
        - condition: time
          after:  '10:00:00'
          before: '15:00:00'
        - condition: state
          entity_id: input_boolean.hasp_devplate_mail_triggered
          state: 'off'
    action:
    - service: input_number.set_value
      data_template:
        entity_id: "input_number.hasp_devplate_mail_activepage"
        value: "{{states('input_number.hasp_devplate_activepage')|int}}"
    - service: input_boolean.turn_on
      data:
        entity_id: input_boolean.hasp_devplate_mail_triggered
    - service: mqtt.publish
      data:
        topic: "hasp/devplate/command/json"
        payload_template: >- 
          [
            "page 1"
          ]

########################################################################################     

  - alias: hasp_devplate_p1_ClearMailTrigger
    trigger:
      - platform: mqtt
        topic: "hasp/devplate/state/p[1].b[4]"
        payload: "ON"
    condition:
      condition: state
      entity_id: input_boolean.hasp_devplate_mail_triggered
      state: 'on'
    action:
    - service: mqtt.publish
      data:
        topic: "hasp/clear_mail_trigger"

########################################################################################     

  - alias: hasp_devplate_p1_ClearMailTriggerHandler
    trigger:
      - platform: mqtt
        topic: "hasp/clear_mail_trigger"
    action:
    - service: input_boolean.turn_off
      entity_id: input_boolean.hasp_devplate_mail_triggered
    - service: mqtt.publish
      data:
        topic: "hasp/devplate/command/page"
        payload_template: " {{ states('input_number.hasp_devplate_mail_activepage')|int }} "

########################################################################################     

  - alias: hasp_devplate_p1_SetMailFlag
    trigger:
    - platform: state
      entity_id: 'binary_sensor.devplate_connected'
      to: 'on'
    - platform: homeassistant
      event: start
    - platform: state
      entity_id: 'input_boolean.hasp_devplate_mail_triggered'
      to: "on"
    condition:
      condition: state
      entity_id: input_boolean.hasp_devplate_mail_triggered
      state: 'on'
    action:
    - service: mqtt.publish
      data:
        topic: "hasp/devplate/command/json"
        payload_template: >- 
          [
            "p[1].b[4].txt=\"Mail's In\"",
            "p[1].b[4].bco=2016"
          ]

########################################################################################     

  - alias: hasp_devplate_p1_ClearMailFlag
    trigger:
    - platform: state
      entity_id: 'binary_sensor.devplate_connected'
      to: 'on'
    - platform: homeassistant
      event: start
    - platform: state
      entity_id: 'input_boolean.hasp_devplate_mail_triggered'
      to: "off"
    condition:
      condition: state
      entity_id: input_boolean.hasp_devplate_mail_triggered
      state: 'off'
    action:
    - service: mqtt.publish
      data:
        topic: "hasp/devplate/command/json"
        payload_template: >- 
          [
            "p[1].b[4].txt=\"       No Mail\"",
            "p[1].b[4].bco={{ states('input_number.hasp_devplate_selectedbackgroundcolor')|int }}"
          ]

##############################################################################
`

**Comment by aderusha** [comment]
I'm not 100% sure I entirely understand the workflow above but I can offer some observations:

* Blueprints are a nice feature which allow me to create generic automations and distribute them such that users can pick their own settings at deploy time.  If you just want HASP to do a thing on your own system, feel free to continue using automations, either via the GUI or through YAML.  Blueprints aren't a requirement unless you want to distribute your automations to other people and to give them a way to customize the result for their own system.
* The binary_sensor object `binary_sensor.devplate_connected` no longer exists in HASPone as it's function is now redundant.  The same functionality is now available via the state of `sensor.devplate_sensor`.
* The `input_boolean` that was being created by the packages no longer exists, as blueprints cannot create helper objects.  I can however use MQTT discovery to create things, so that same function now exists but the entity is called `number.devplate_active_page`

**Reply by eboston** [reply]
This was from the pre-HASPone automation.  After posting, I thought more about it and figure it probably will be best as an external automation.

What is happening is I have a sensor in my mailbox that detects when the door is opened (mail being delivered).  It sets a MQTT topic that triggers a Node-Red flow that causes that HASP to record the current page it is on, change to page 1, change the text on button 4 to "Mail's In" with a green background.  The Node-Red sends these commands to the plate group so all HASPs get it.

When button 4 is pressed on any of the HASPs, a MQTT message is sent to the group saying to clear the mail flag.  The HASPs reset the button 4 text and then returns to the previous page.

Probably the only thing that could be handled with the existing system is clearing the flag.  When the button is pressed, and action is performed.  Everything else would need to be done with regular automation.

Is that clear as mud now?  :)

**Reply by aderusha** [reply]
I totally get it and also that is FRICKEN AWESOME and I want to do that too.  What are you using for your mailbox sensor, and do you like it?

**Reply by eboston** [reply]
I used a D1 mini with a reed switch.  When the mailbox door opens, it powers off the D1.  When the door is closed again, it powers up.  The code checks the reason for the last reset.  If it was a power up, it sends the message.  The D1 goes into deep sleep mode for a minute and when it wakes up, sends status.

**Reply by eboston** [reply]
So I have been able to get it working, sort of, by creating the automation manually.  Since the active page value doesn't exist anymore, it doesn't go back to the page it was on.

To use the new value, do I just change 'input_number.hasp_desktop_activepage' to 'number.desktop_activepage' ?

**Reply by wr420** [reply]
Curious why you have the reed switch turn the D1 off when it opens rather than monitoring the state of the reed switch?
Running on battery? But in that case would it probably be better to leave it off and turn it on when the door was opened.
Sorry, I'm playing with a similar setup for a more generic use so just wondering if I'm missing something.


Have you looked at these? I have a few and they could work on a mailbox door.  Basically a ball tilt [switch](https://www.amazon.com/gp/product/B07TF6XT29/ref=ppx_yo_dt_b_asin_title_o03_s00?ie=UTF8&psc=1)

**Reply by aderusha** [reply]
One thing I'll note about ball tilt switches - if you live in a cold climate, condensation can get inside and freeze up the ball so it no longer moves reliably.  I had a similar setup on my garage doors w/ a cheap Z-Wave tilt sensor, and replaces the ball switches with mercury tilt switches.  Cheap, drop-in replacement that has no problems with Michigan weather!

**Reply by eboston** [reply]
On the 8266, the interrupt handler for a GPIO pin and deep sleep wake up is the same so you can't tell which one triggered it.  I put the 8266 into deep sleep for an hour and have it wake up to send status information.  If the door was on a GPIO, I wouldn't be able to tell the difference between a door opening and waking from deep sleep.  As for turning it on when the door is opened, it may not be opened long enough for the board to startup, connect, and send the status before it closes again.

With the ESP32, you can get that information and I have code to support it but have been having problems finding the right hardware.  I want an ESP32 board with both a battery charger and external antenna.  I can find boards with one or the other, but not both.

I used the reed switch so I didn't have to have wires running to door.  I looked at mercury and ball switches too.  Reed just worked out better for me is all.

**Comment by wr420** [comment]
Ah, good to know.  I’m in California so the weather is usually more cooperative, but water ingress , freezing or not, could cause problems in any climate.
Thanks for the heads up.  I was originally looking for mercury switches but I think I figured “mercury” probably illegal to sell in California anyway. lol!

**Comment by wr420** [comment]
Thanks for the heads up.  You are clearly a few level up on me😉

Very cleaver solution.

















On the 8266, the interrupt handler for a GPIO pin and deep sleep wake up is the same so you can't tell which one triggered it. I put the 8266 into deep sleep for an hour and have it wake up to send status information. If the door was on a GPIO, I wouldn't be able to tell the difference between a door opening and waking from deep sleep. As for turning it on when the door is opened, it may not be opened long enough for the board to startup, connect, and send the status before it closes again.

With the ESP32, you can get that information and I have code to support it but have been having problems finding the right hardware. I want an ESP32 board with both a battery charger and external antenna. I can find boards with one or the other, but not both.

I used the reed switch so I didn't have to have wires running to door. I looked at mercury and ball switches too. Reed just worked out better for me is all.
